plugins/inputs/snmp2/snmp.go

```diff
@@ -0,0 +1,624 @@
+	return s
+}
+
+// NestedError wraps an error returned from deeper in the code.
+type NestedError struct {
+	// Err is the error from where the NestedError was constructed.
+	Err error
+	// NestedError is the error that was passed back from the called function.
+	NestedErr error
+}
+
+// Error returns a concatenated string of all the nested errors.
+func (ne NestedError) Error() string {
+	return ne.Err.Error() + ": " + ne.NestedErr.Error()
+}
+
+// Errorf is a convenience function for constructing a NestedError.
+func Errorf(err error, msg string, format ...interface{}) error {
+	return NestedError{
+		NestedErr: err,
+		Err:       fmt.Errorf(msg, format...),
+	}
+}
+
+func init() {
+	inputs.Add("snmp2", func() telegraf.Input {
+		return &Snmp2{
+			Retries:        5,
+			MaxRepetitions: 50,
+		}
+	})
+}
+
+// SampleConfig returns the default configuration of the input.
+func (s *Snmp2) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns a one-sentence description on the input.
+func (s *Snmp2) Description() string {
+	return description
+}
+
+// Gather retrieves all the configured fields and tables.
+// Any error encountered does not halt the process. The errors are accumulated
+// and returned at the end.
+func (s *Snmp2) Gather(acc telegraf.Accumulator) error {
+	var errs Errors
+	for _, agent := range s.Agents {
+		gs, err := s.getConnection(agent)
+		if err != nil {
+			errs = append(errs, Errorf(err, "agent %s", agent))
+			continue
+		}
+
+		// First is the top-level fields. We treat the fields as table prefixes with an empty index.
+		t := Table{
+			Name:   s.Name,
+			Fields: s.Fields,
+		}
+		topTags := map[string]string{}
+		if err := s.gatherTable(acc, gs, t, topTags, false); err != nil {
+			errs = append(errs, Errorf(err, "agent %s", agent))
+		}
+
+		// Now is the real tables.
+		for _, t := range s.Tables {
+			if err := s.gatherTable(acc, gs, t, topTags, true); err != nil {
+				errs = append(errs, Errorf(err, "agent %s", agent))
+			}
+		}
+	}
+
+	if errs == nil {
+		return nil
+	}
+	return errs
+}
+
+func (s *Snmp2) gatherTable(acc telegraf.Accumulator, gs snmpConnection, t Table, topTags map[string]string, walk bool) error {
+	rt, err := t.Build(gs, walk)
+	if err != nil {
+		return err
+	}
+
+	for _, tr := range rt.Rows {
+		if !walk {
+			// top-level table. Add tags to topTags.
+			for k, v := range tr.Tags {
+				topTags[k] = v
+			}
+		} else {
+			// real table. Inherit any specified tags.
+			for _, k := range t.InheritTags {
+				if v, ok := topTags[k]; ok {
+					tr.Tags[k] = v
+				}
+			}
+		}
+		if _, ok := tr.Tags["agent_host"]; !ok {
+			tr.Tags["agent_host"] = gs.Host()
+		}
+		acc.AddFields(rt.Name, tr.Fields, tr.Tags, rt.Time)
+	}
+
+	return nil
+}
+
+// Build retrieves all the fields specified in the table and constructs the RTable.
+func (t Table) Build(gs snmpConnection, walk bool) (*RTable, error) {
+	rows := map[string]RTableRow{}
+
+	tagCount := 0
+	for _, f := range t.Fields {
```

Comment:
here I think we can check that the OID is a valid table field with something like:

```
if walk && strings.HasSuffix(f.Oid, ".0") {
    // log error and continue?
}
```